File: backend/app/Models/WalletRechargeRequest.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class WalletRechargeRequest extends Model
{
    protected $fillable = [
        'user_id',
        'amount',
        'currency',
        'receipt_image',
        'payment_method',
        'bank_name',
        'transaction_reference',
        'notes',
        'status',
        'admin_notes',
        'processed_by',
        'processed_at',
    ];

    protected $casts = [
        'amount' => 'decimal:2',
        'processed_at' => 'datetime',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    protected $appends = ['status_text', 'formatted_amount'];

    // Status constants
    const STATUS_PENDING = 'pending';
    const STATUS_APPROVED = 'approved';
    const STATUS_REJECTED = 'rejected';

    // Relationships
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class, 'user_id', 'id');
    }

    public function processedBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'processed_by', 'id');
    }

    // Accessors
    public function getStatusTextAttribute(): string
    {
        return match($this->status) {
            self::STATUS_PENDING => 'قيد الانتظار',
            self::STATUS_APPROVED => 'مقبول',
            self::STATUS_REJECTED => 'مرفوض',
            default => $this->status,
        };
    }

    public function getFormattedAmountAttribute(): string
    {
        return number_format($this->amount, 2) . ' ' . $this->currency;
    }

    public function getReceiptUrlAttribute(): ?string
    {
        if (!$this->receipt_image) {
            return null;
        }

        // If it's already a full URL, return it
        if (filter_var($this->receipt_image, FILTER_VALIDATE_URL)) {
            return $this->receipt_image;
        }

        // Otherwise, prepend storage URL
        return url('storage/' . $this->receipt_image);
    }

    // Methods
    public function approve(string $adminId, ?string $adminNotes = null): bool
    {
        if ($this->status !== self::STATUS_PENDING) {
            return false;
        }

        try {
            \DB::beginTransaction();

            // Update request status
            $this->status = self::STATUS_APPROVED;
            $this->admin_notes = $adminNotes;
            $this->processed_by = $adminId;
            $this->processed_at = now();
            $this->save();

            // Credit user wallet
            $wallet = Wallet::where('user_id', $this->user_id)->first();
            if (!$wallet) {
                $wallet = Wallet::createForUser($this->user_id);
            }

            $wallet->credit(
                $this->amount,
                'إعادة شحن المحفظة - طلب رقم ' . $this->id,
                'recharge_request_' . $this->id,
                [
                    'recharge_request_id' => $this->id,
                    'payment_method' => $this->payment_method,
                    'bank_name' => $this->bank_name,
                ]
            );

            \DB::commit();
            return true;
        } catch (\Exception $e) {
            \DB::rollBack();
            \Log::error('Failed to approve recharge request: ' . $e->getMessage());
            return false;
        }
    }

    public function reject(string $adminId, ?string $adminNotes = null): bool
    {
        if ($this->status !== self::STATUS_PENDING) {
            return false;
        }

        $this->status = self::STATUS_REJECTED;
        $this->admin_notes = $adminNotes;
        $this->processed_by = $adminId;
        $this->processed_at = now();

        return $this->save();
    }

    public function isPending(): bool
    {
        return $this->status === self::STATUS_PENDING;
    }

    public function isApproved(): bool
    {
        return $this->status === self::STATUS_APPROVED;
    }

    public function isRejected(): bool
    {
        return $this->status === self::STATUS_REJECTED;
    }

    // Scopes
    public function scopePending($query)
    {
        return $query->where('status', self::STATUS_PENDING);
    }

    public function scopeApproved($query)
    {
        return $query->where('status', self::STATUS_APPROVED);
    }

    public function scopeRejected($query)
    {
        return $query->where('status', self::STATUS_REJECTED);
    }

    public function scopeRecent($query)
    {
        return $query->orderBy('created_at', 'desc');
    }
}

File: lib/models/wallet_recharge_request.dart
class WalletRechargeRequest {
  final int id;
  final String? referenceNumber;
  final String userId;
  final double amount;
  final String currency;
  final String? receiptUrl;
  final String? receiptImagePath; // Local path for receipt image
  final String? paymentMethod;
  final String? bankName;
  final String? transactionReference;
  final String? notes;
  final String status;
  final String statusText;
  final String? adminNotes;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final DateTime? processedAt;

  WalletRechargeRequest({
    required this.id,
    this.referenceNumber,
    required this.userId,
    required this.amount,
    required this.currency,
    this.receiptUrl,
    this.receiptImagePath,
    this.paymentMethod,
    this.bankName,
    this.transactionReference,
    this.notes,
    required this.status,
    String? statusText,
    this.adminNotes,
    required this.createdAt,
    this.updatedAt,
    this.processedAt,
  }) : statusText = statusText ?? _getStatusText(status);

  static String _getStatusText(String status) {
    switch (status) {
      case 'pending':
        return 'قيد الانتظار';
      case 'approved':
        return 'تمت الموافقة';
      case 'rejected':
        return 'مرفوض';
      default:
        return 'غير معروف';
    }
  }

  factory WalletRechargeRequest.fromJson(Map<String, dynamic> json) {
    return WalletRechargeRequest(
      id: json['id'] is int ? json['id'] : int.tryParse(json['id'].toString()) ?? 0,
      referenceNumber: json['reference_number'] as String?,
      userId: json['user_id'] as String? ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      currency: json['currency'] as String? ?? 'SAR',
      receiptUrl: json['receipt_url'] as String?,
      receiptImagePath: json['receipt_image_path'] as String?,
      paymentMethod: json['payment_method'] as String?,
      bankName: json['bank_name'] as String?,
      transactionReference: json['transaction_reference'] as String?,
      notes: json['notes'] as String?,
      status: json['status'] as String? ?? 'pending',
      statusText: json['status_text'] as String?,
      adminNotes: json['admin_notes'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
      processedAt: json['processed_at'] != null
          ? DateTime.parse(json['processed_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'reference_number': referenceNumber,
      'user_id': userId,
      'amount': amount,
      'currency': currency,
      'receipt_url': receiptUrl,
      'receipt_image_path': receiptImagePath,
      'payment_method': paymentMethod,
      'bank_name': bankName,
      'transaction_reference': transactionReference,
      'notes': notes,
      'status': status,
      'status_text': statusText,
      'admin_notes': adminNotes,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
      'processed_at': processedAt?.toIso8601String(),
    };
  }

  bool get isPending => status == 'pending';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';

  String get formattedAmount => '${amount.toStringAsFixed(2)} $currency';
}
